clang/lib/AST/ASTContext.cpp

```diff
@@ -3830,6 +3830,16 @@ QualType ASTContext::getDecayedType(QualType T) const {
   return getDecayedType(T, Decayed);
 }
 
+QualType ASTContext::getConstantArrayFromArrayParameterType(QualType Ty) const {
+  if (Ty->isConstantArrayType() && !Ty->isArrayParameterType())
+    return Ty;
```

Comment:
The name of this method implies that it would be wrong to call this with anything other than an `ArrayParameterType`.

Maybe it makes sense to make this a method on the ArrayParameterType so that you can operate on `this` knowing it is the correct type?